dashboard/dashboard.go

```diff
@@ -0,0 +1,340 @@
+func (db *dashboard) Start(server *p2p.Server) error {
+	//log.Trace("Start() called", "config", db.config)
+
+	go db.collectData()
+
+	http.HandleFunc("/", db.webHandler)
+	http.Handle("/api", websocket.Handler(db.apiHandler))
+
+	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", db.config.Host, db.config.Port))
+	if err != nil {
+		return err
+	}
+	db.listener = listener
+
+	go func() {
+		//log.Trace("Starting server...")
+
+		if err := http.Serve(listener, nil); err != nil {
+			log.Warn("Server failed", "err", err)
+		}
+	}()
+
+	return nil
+}
+
+// Stop implements node.Service, stopping the data collection thread and the connection listener of the dashboard
+func (db *dashboard) Stop() error {
+	//log.Trace("Terminating dashboard...")
+
+	db.lock.Lock()
+	defer db.lock.Unlock()
+
+	var err error
+	// Close the connection listener
+	if err = db.listener.Close(); err != nil {
+		log.Warn("Failed to close listener", "err", err)
+	}
+
+	close(db.quit) // Notifies collectData and apiHandler
+
+	for _, c := range db.conns {
+		if err := c.conn.Close(); err != nil {
+			c.logger.Warn("Failed to close connection", "err", err)
+		}
+	}
+	db.conns = db.conns[:0]
+
+	return err
+}
+
+// webHandler handles all non-api requests, simply flattening and returning the dashboard website.
+func (db *dashboard) webHandler(w http.ResponseWriter, r *http.Request) {
+	//log.Trace("webHandler() called", "r.URL", r.URL)
+	log.Info("Request", "URL", r.URL)
+
+	path := r.URL.String()
+
+	// If the path of the assets is manually set
+	if db.config.Assets != "" {
+		// Create the filename for ReadFile
+		var buffer bytes.Buffer
+		buffer.WriteString(db.config.Assets)
+		switch path {
+		case "/":
+			buffer.WriteString("/dashboard.html")
+		default:
+			buffer.WriteString(path)
+		}
+
+		file, err := ioutil.ReadFile(buffer.String())
+		if err != nil {
+			// TODO (kurkomisi): Warn or Crit?
+			log.Warn("Failed to read file", "err", err)
+			// TODO (kurkomisi): Should I inform the client?
+			return
+		}
+		w.Write(file)
+		return
+	}
+
+	switch path {
+	case "/":
+		index, err := Asset("dashboard.html")
+		if err != nil {
+			log.Warn("Failed to load the index", "err", err)
+			return
+		}
+		w.Write(index)
+	case "/js/handlers.js":
+		tmpl, err := Asset("js/handlers.js")
+		if err != nil {
+			log.Warn("Failed to load the asset", "path", path, "err", err)
+			return
+		}
+		handlers := new(bytes.Buffer)
+		// TODO (kurkomisi): Save the generated template to avoid the repeated generation?
+		// set the sample limits for the client
+		if err = template.Must(template.New("").Parse(string(tmpl))).Execute(handlers, map[string]interface{}{
+			"processorSampleLimit": processorSampleLimit,
+			"memorySampleLimit":    memorySampleLimit,
+			"trafficSampleLimit":   trafficSampleLimit,
+		}); err != nil {
+			log.Warn("Failed to render the dashboard handlers template", "err", err)
+			return
+		}
+		w.Write(handlers.Bytes())
```

Comment:
I don't think it's a good idea to use templates here (or anywhere to be honest). You have a bi-directional websocket connection between the server and the client. Any data that you need should be sent over the socket, not via templating. This just makes the logic harder to follow. 

More generally, there websites and webapps are two different concepts. Websites are generated server side and the final content sent to the client. Webapps on the other hand send the necessary static code to the client, and the client will assemble the final content based on the code and any other relevant data it gets or retrieves. In this perspective, the dashboard is a web application, it generates all its views client side, not server side.

As such, please remove all templating from the dashboard backend. The backend shouldn't need to care about these kind of things.